components/base/kinematicbase/differentialDrive.go

```diff
@@ -82,15 +91,23 @@ type differentialDriveKinematics struct {
 	logger                        logging.Logger
 	planningFrame, executionFrame referenceframe.Model
 	options                       Options
+	noLocalizerCacheInputs        []referenceframe.Input
+	mutex                         sync.RWMutex
 }
 
 func (ddk *differentialDriveKinematics) Kinematics() referenceframe.Frame {
 	return ddk.planningFrame
 }
 
 func (ddk *differentialDriveKinematics) CurrentInputs(ctx context.Context) ([]referenceframe.Input, error) {
+	// If no localizer is present, CurrentInputs returns the expected position of the robot assuming after
+	// each part of move command was completed accurately.
```

Comment:
So this ties in explore mode to the differential base wrapper for now, is there a way to catch inputs directly in the explore mode of nav service?

That would be a cleaner design. Imagine a Frankensteinish future where someone writes a nav module for explore but uses our motion unwittingly did not pass in a localizer as a movement sensor for their algorithm, but uses it within the module code as well. This seems like it would be a way to overwrite the inputs from that movement sensor in the module.

Reply:
This would be cleaner, but would require a redesign to the differential base or PlanMotion. Currently, the motion plan is fed, step by step to the differential base. Therefore we can monitor which steps it has completed and which are still to be done.

But in differential base, each step gets decomposed to multiple substeps. When you have a Localizer, this is used to allow the differential base to handle when movement actions do not result in the desired goal location for that step. If you want to move to (x,y) but after performing the move that should get you there, you end up at (x1,y1). The differential base handles the correction and subsequent move actions to get you fully to (x,y). 

For non-Localizer-based differential drives, we ASSUME the goal has been reached accurately, however, each step of the motion plan is still decomposed into a spin and move straight command. This caching of the currentInputs is used to allow the navigation service to know when each sub-step has been complete as it can not ping the Localizer directly.

One potential solution this is to attempt to decompose the steps into their substeps in navigation. Instead of having a motion plan that consist of [`[0,0,0]`, `[x,y,theta]`] (note: this is always what the motion plan looks like as the world state is empty so a direct line can be created from current position to goal position) we would create a motion plan [`[0,0,0]`, `[0,0,theta]` `[x,y,theta]`]. splitting up the move action into a spin and move straight action from the start.

The reason I proposed the caching solution is that this allows no changes to be made to PlanMotion, nor changes inside the motion explore service that would involve editing the motion plan. Instead, we are just creating a fake localizer. I'd be happy to discuss changing this, especially if there are other reasons we may want to move this capability out of the differential base and/or change our current patterns for motion planning.

Reply:
Are we doing the refactor to make it cleaner in this pr or in a follow-up ticket that we backlog?

Reply:
Currently, differential_drive is not heavily used. I dont think it's worth the time to refactor, I'd prefer to handle any changes to this noLocalizer flow when there is a bigger move to update differential_drive. But I'm open to alterative opinions